pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,219 @@
+	}
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer(),
+		).
+		ResyncEvery(1*time.Minute). // an arbitrary number, not too high not too low
+		ToController("MachineDeletionHooksController", eventRecorder.WithComponentSuffix("cluster-machine-deletion-hooks-controller"))
+}
+
+func (c *machineDeletionHooksController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	// stop if the machine API is not functional
+	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	var errs []error
+	if err := c.addDeletionHookToMasterMachines(ctx); err != nil {
+		errs = append(errs, err)
+	}
+	// attempt to remove hooks from machines pending deletion and without running etcd member
+	if err := c.attemptToDeleteMachineDeletionHook(ctx, syncCtx.Recorder()); err != nil {
+		errs = append(errs, err)
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+// it will only remove the hook once it has identified that the Machine resource is being deleted and it doesn't host an etcd member
+func (c *machineDeletionHooksController) attemptToDeleteMachineDeletionHook(ctx context.Context, recorder events.Recorder) error {
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	// machines with the deletion hooks are the ones that host etcd members
+	machinesWithHooks := ceohelpers.FilterMachinesWithMachineDeletionHook(masterMachines)
+
+	// if at least one member machine is pending deletion
+	machinesPendingDeletion := ceohelpers.FilterMachinesPendingDeletion(machinesWithHooks)
+	if len(machinesPendingDeletion) == 0 {
+		klog.V(4).Infof("currently, we don't have a machine pending deletion")
+		return nil
+	}
+
+	// get the members, issue a live call instead of reading from the config map
+	// because we need to take into account learners
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+	currentMemberIPListSet := sets.NewString()
+	for _, member := range members {
+		memberIP, err := ceohelpers.MemberToNodeInternalIP(member)
+		if err != nil {
+			return err
+		}
+		currentMemberIPListSet.Insert(memberIP)
+	}
+
+	// find a machine that doesn't have a member
+	for machinePendingDeletionIP, machinePendingDeletion := range indexMachinesByNodeInternalIP(machinesPendingDeletion) {
+		if currentMemberIPListSet.Has(machinePendingDeletionIP) {
+			// Ignore machines that still have a member present
+			continue
+		}
+		machinePendingDeletionCopy := machinePendingDeletion.DeepCopy()
+		removeMachineDeletionHook(machinePendingDeletionCopy)
+		if _, err := c.machineClient.Update(ctx, machinePendingDeletionCopy, metav1.UpdateOptions{}); err == nil {
```

Comment:
Aggregate these errors so you can get as right as you can if something is wonky with the API.